### PR TITLE
add business type extractor

### PR DESCRIPTION
**Pull request by ZwEin27**

### CODE REVIEW

File: digBusinessTypeExtractor/documentation_classification.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+DATA_DIR = os.path.join(os.path.dirname(__file__), 'data')
+
+class WEDC(object):
+
+    ##################################################################
+    # Basic Methods
+    ##################################################################
+
+    def __init__(self, data_path=os.path.join(DATA_DIR, 'dataset.csv'), vectorizer_type='count', classifier_type='knn'):
```
Comment (jasonslepicka) [inline digBusinessTypeExtractor/documentation_classification.py]: There some serious problems with this approach.  If we have to load the dataset and train it every time we run the program it's going to be really slow to start up.  We also need to update setup.py to including the training resources. If we're loading it and training it every time we call predict, it's going to be even slower.  We should allow users to pass in a WEDC in the BusinessTypeExtractor.  We should also add a method to the WEDC utility to let users serialize their trained classifers. 

Predict should just include

```
vectors = self.vectorizer.fit_transform(test_data_corpus).toarray()
test_X = vectors
pred_y = self.classifier.predict(test_X)
pred_y = [DC_CATEGORY_NAMES[int(i)] for i in pred_y]
return pred_y
```

**Review by jasonslepicka — CHANGES_REQUESTED** [review]
We need to refactor the WEDC and BusinessTypeExtractor so we aren't building and training the model on every call to predict.

File: digBusinessTypeExtractor/documentation_classification.py (hunk 1 above)
Comment (ZwEin27) [reply digBusinessTypeExtractor/documentation_classification.py]: Hi Jason, thanks for pointing out, :). I have updated it, and please have a look for the detail, thank you

File: digBusinessTypeExtractor/documentation_classification.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        if not vectorizer_model_path:
+            vectorizer_model_path = self.vectorizer_model_path
+
+        data_corpus = [data] if isinstance(data, basestring) else data
+
+        # load trained classifier and vectorizer models
+        classifier = vectorizer = None
+        try:
+            classifier = joblib.load(classifier_model_path) 
+            vectorizer = joblib.load(vectorizer_model_path)
```
Comment (jasonslepicka) [inline digBusinessTypeExtractor/documentation_classification.py]: It looks like you're loading the classifier and vectorizer every time you call predict.  You should test to see if classifier and vectorizer are None before loading them again.

File: digBusinessTypeExtractor/documentation_classification.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    def load_data(self, filepath=None):
+
+        def clean_content(content):
+            content = content.lower()
+
+            # remove digits and punctuations
+            content = ''.join([_ for _ in content if (_ >= 'a' and _ <= 'z') or (_ in ' \t')])
+            
+            # remove stopwords
+            # content = ' '.join([_ for _ in content.split() if _ not in STOP_WORDS])
```
Comment (jasonslepicka) [inline digBusinessTypeExtractor/documentation_classification.py:112]: It looks like you aren't removing stop words.  It looks like we can use "english" or just pass STOP_WORDS to TfIdfVectorizer (you might have to convert it to a list first though).  http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

File: digBusinessTypeExtractor/documentation_classification.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+        # full matrix
+        # vectors = self.vectorizer.fit_transform(self.corpus).toarray()
+        # self.classifier.fit(vectors, self.labels)
+        
+        # sparse matrix
+        vectors = self.vectorizer.fit_transform(self.corpus)
+        self.classifier.fit(vectors, self.labels)
+
+        joblib.dump(self.classifier, self.classifier_model_path)
```
Comment (jasonslepicka) [inline digBusinessTypeExtractor/documentation_classification.py]: I don't think you should write the model out by default.  This should be another method that the user should be able to call after training.

File: digBusinessTypeExtractor/documentation_classification.py (hunk 2 above)
Comment (jasonslepicka) [reply digBusinessTypeExtractor/documentation_classification.py]: Why aren't you using the self.classifier and self.vectorizer?

File: digBusinessTypeExtractor/documentation_classification.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    'unknown': -1,
+    'others': 1,
+    'massage': 2,
+    'escort': 3,
+    'job_ads': 4
+}
+
+USER_HOME_PATH = expanduser("~")
+
+APP_DIR = os.path.join(USER_HOME_PATH, 'WEDC_DATA')
```
Comment (jasonslepicka) [inline digBusinessTypeExtractor/documentation_classification.py:48]: I don't think it's a good idea to create a default directory, especially not in the user's home.  You may not always have access to it.  When we're running on the cluster, some users don't even have a home directory.  The user should always specify where it is or at least make it an environment variable so we can override it.  I don't think having init_env() is a good idea.  if we're going to ship the package with the data, we might as well include a default tfidf/cosine model in the package as well, so we don't have to train the model everytime we run the program on a new machine.

**Review by jasonslepicka — CHANGES_REQUESTED** [review]
Still need some more changes.  Sorry :/

File: digBusinessTypeExtractor/documentation_classification.py (hunk 3 above)
Comment (ZwEin27) [reply digBusinessTypeExtractor/documentation_classification.py:112]: done for stopwords

File: digBusinessTypeExtractor/documentation_classification.py (hunk 4 above)
Comment (ZwEin27) [reply digBusinessTypeExtractor/documentation_classification.py]: updated logic to save model

File: digBusinessTypeExtractor/documentation_classification.py (hunk 2 above)
Comment (ZwEin27) [reply digBusinessTypeExtractor/documentation_classification.py]: update logic for self.classifier and self.vectorizer

**Review by jasonslepicka — APPROVED** [review]

**Pull request merged**